### DISCUSSION TITLE
using homebrew provided github actions for generating bottles

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2006] [Tap maintenance and Homebrew development]
i've been working on updating the _homebrew-freecad_ tap to use github actions most notably the standard `tests.yml` and the `publish.yml` when creating a new tap / repo. i feel like have a somewhat decent understanding of what i'm doing but still have some questions that i do not know the answers too.

1. i created [simple dummy repo][1] for experimenting with homebrew and github actions

it's been great to mess around with github actions and homebrew more from a tap maintainer perspective and not blowing up an existing repo 🤞.

2. i checked in one of the formula file from the _homebrew-freecad_ tap ie. swig and after some monkeying around i was able to build a bottle for all the 3 most recent major releases of macos.
  - mojave
  - catalina
  - bigsur

i was able to build bottles for these OS's using virtual machines i setup an extra macbook i have. and have each VM acting as a _self-hosted_ runner as github only provides a catalina build environment.

so using the `tests.yml` along with `brew test-bot` i built swig for all mentioned OS's. and now i checked in another file ie. python and am experimenting with it.

however there are **2** things that are presently blocking me from publishing my python bottles.

1. if i create a PR ie. create branch, add the `python.rb` within my new branch. then open a PR against my pg repo/homebrew tap. i can open a PR and fire off the jobs to be run on my _self-hosted_ runners.

however, i'm presently seeing [this issue][2], which is something i didn't encounter when working with the swig formula.

it seems the existing homebrew install on these virtual machines is causing issues for these _self-hosted_ runners.

i've worked around this issue by uninstalling homebrew, and then reinstalling hombrew globally ie. within `/usr/local` using the same user that i'm using to start the `run.sh` github actions runner script. however, after running through one action / workflow, and then attempting another i begin to see those errors mentioned the log, ie.

```
Uncommitted files:
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-performance-1.11.4/config/default.yml
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-performance-1.11.4/config/obsoletion.yml
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-performance-1.11.4/lib/rubocop-performance.rb
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-performance-...
omitted the rest for brevity.

Error: 1 failed step!
brew doctor
Error: Process completed with exit code 1.
```

---

and the **second** 2nd issue i've run into, is, if i have a PR open, and then submit a follow commit, ie. i opened a PR creating the python formulae file for my custom homebrew tap, and then i decide to update amend the original file i checked in, so my PR is composed of 2 commit now, the `publish.yml` action will fail due to a `git cherry-pick` issue with not being able to merge the commits.

[this error][3] happens after the bottles have been successfully built and then the `pr-pull` label is added to the PR request to publish the bottles.

**is this the intended behavior of this publish** workflow? or is there a more elegant way to handle these things?

any & all help is greatly appreciated. 🙏

[1]: <https://github.com/ipatch/homebrew-freecad-pg13>
[2]: <https://github.com/ipatch/homebrew-freecad-pg13/runs/3366723145#step:6:47>
[3]: <https://github.com/ipatch/homebrew-freecad-pg13/runs/3366523919?check_suite_focus=true#step:4:19>

**Comment by dawidd6** [comment]
Your use case isn't really what we aim for with our `brew tap-new` generated workflows. The setup that you want/have is more like what we already have in homebrew/core. I guess you probably want to run cleanups before building bottles and after that. And since you already have "before" cleanup, it might be a good idea to also run "after" cleanup at the end of workflow:
```yaml
      - name: Post cleanup
        if: always()
        run: brew test-bot --only-cleanup-after
```

Also setting `GITHUB_ACTIONS_HOMEBREW_MACOS_SELF_HOSTED` env variable could help here.

**Reply by ipatch** [reply]
what benefit does setting

```
GITHUB_ACTIONS_HOMEBREW_MACOS_SELF_HOSTED
```

provide?

**Reply by dawidd6** [reply]
Can't tell for sure, would need to check the code here and there. It does some extra magic for macOS self hosted runners.
